### ISSUE TITLE
Trying to claim a badge in a group deletes the group

### ISSUE THREAD

**Issue by kitsuta** [magfest/ubersystem#1119]
I do not yet know exactly what causes it, but when you try to use register_group_member to claim a badge in a group, the group is immediately deleted and throws a 500 error at you for trying to load its group_members page (since the ID is no longer associated with anything).

This bug is live in Magstock so we need to fix and deploy this ASAP - I know we had admins who were using groups to invite bands so those links may have gone out already.

**Comment by EliAndrewC** [comment]
WHOOPS!

As you can probably tell, I didn't do all that much testing on the group stuff, since I didn't think it was going to get used for MAGStock.  Since it is, I should probably go through and do a more thorough round of testing, which I figured I could put off until getting ready to launch prereg for MagClassic :/

**Comment by kitsuta** [comment]
Well the good news is that according to the person who was adding groups, none of these links have gone out yet. Phew!

This still needs to be fixed pretty soon. The puzzling thing is that when you comment out `session.delete_from_group(badge_being_claimed, group)` on line 310, the behavior is fixed... but nothing in that function has changed. This is something further back - possibly there's a cascade somewhere in the DB architecture?

**Comment by EliAndrewC** [comment]
Hmm, I did reverse the directionality of the ``relationship`` declarations.  Although my intention was to keep the functionality exactly the same as it was before, it's possible that while doing that I accidentally made a typo or logic mistake that added a cascading delete where one didn't exist before.

**Comment by kitsuta** [comment]
Looks like you explicitly added `cascade='all'` to the `leader` relationship in Group class and the `group` relationship of the Attendee class. This looks like the culprit, but after removing this options and restarting the server, I get this error:

```
500 Internal Server Error

The server encountered an unexpected condition which prevented it from fulfilling the request.

Traceback (most recent call last):
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/CherryPy-3.6.0-py3.4.egg/cherrypy/_cprequest.py", line 670, in respond
    response.body = self.handler()
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/CherryPy-3.6.0-py3.4.egg/cherrypy/lib/encoding.py", line 217, in __call__
    self.body = self.oldhandler(*args, **kwargs)
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/CherryPy-3.6.0-py3.4.egg/cherrypy/_cpdispatch.py", line 61, in __call__
    return self.callable(*self.args, **self.kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 159, in with_timing
    return func(*args, **kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 150, in with_caching
    return func(*args, **kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 174, in with_session
    retval = func(*args, session=session, **kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 258, in with_restrictions
    return func(*args, **kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/decorators.py", line 220, in with_rendering
    result = func(*args, **kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/site_sections/preregistration.py", line 49, in wrapped
    return func(self, *args, **kwargs)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/site_sections/preregistration.py", line 287, in register_group_member
    group = session.group(group_id)
  File "/home/vagrant/uber/sideboard/plugins/uber/uber/models.py", line 1808, in getter
    return self.query(model).filter_by(id=params).one()
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/SQLAlchemy-1.0.0-py3.4-linux-i686.egg/sqlalchemy/orm/session.py", line 1249, in query
    return self._query_cls(entities, self, **kwargs)
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/SQLAlchemy-1.0.0-py3.4-linux-i686.egg/sqlalchemy/orm/query.py", line 110, in __init__
    self._set_entities(entities)
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/SQLAlchemy-1.0.0-py3.4-linux-i686.egg/sqlalchemy/orm/query.py", line 120, in _set_entities
    self._set_entity_selectables(self._entities)
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/SQLAlchemy-1.0.0-py3.4-linux-i686.egg/sqlalchemy/orm/query.py", line 150, in _set_entity_selectables
    ent.setup_entity(*d[entity])
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/SQLAlchemy-1.0.0-py3.4-linux-i686.egg/sqlalchemy/orm/query.py", line 3184, in setup_entity
    self._with_polymorphic = ext_info.with_polymorphic_mappers
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/SQLAlchemy-1.0.0-py3.4-linux-i686.egg/sqlalchemy/util/langhelpers.py", line 747, in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/SQLAlchemy-1.0.0-py3.4-linux-i686.egg/sqlalchemy/orm/mapper.py", line 1889, in _with_polymorphic_mappers
    configure_mappers()
  File "/home/vagrant/uber/sideboard/env/lib/python3.4/site-packages/SQLAlchemy-1.0.0-py3.4-linux-i686.egg/sqlalchemy/orm/mapper.py", line 2680, in configure_mappers
    raise e
sqlalchemy.exc.InvalidRequestError: One or more mappers failed to initialize - can't proceed with initialization of other mappers.  Original exception was: On Attendee.group, delete-orphan cascade is not supported on a many-to-many or many-to-one relationship when single_parent is not set.   Set single_parent=True on the relationship().
```

**Comment by kitsuta** [comment]
Oh, now here's something I didn't notice - it's _only_ deleting the group, and leaving the attendees inside it untouched.

**Comment by binary1230** [comment]
quick FYI too, groups were only being used (to my knowledge) just to add bands into the system so they could add their comps and stuff.  so the good news is at least the exposure with this bug isn't that huge. still worth fixing : )  god speed.

**Comment by EliAndrewC** [comment]
> Oh, now here's something I didn't notice until now - it's only deleting the group, and leaving the attendees inside it untouched.

Haha, double whoops?  Or maybe only single whoops since we don't generally want badges deleted for some reason.  I guess we can just agree it's a positive number of whoopses.

Starting on this now.

**Comment by kitsuta** [comment]
Fixed in PR #1121.
